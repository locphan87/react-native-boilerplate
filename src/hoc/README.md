# Higher order components

A higher-order component ([HOC](https://reactjs.org/docs/higher-order-components.html)) is an advanced technique in React for reusing component logic.

Concretely, a higher-order component is a function that takes a component and returns a new component.

```js
const EnhancedComponent = higherOrderComponent(WrappedComponent);
```

## Code structure

Naming convention:
* HOC `<HOC_NAME>.hoc.js`
* Test `<HOC_NAME>.hoc.test.js`

```
./src/hoc/withUpdating
├── __snapshots__
│   └── withUpdating.hoc.test.js.snap
├── withUpdating.hoc.js
└── withUpdating.hoc.test.js
```
