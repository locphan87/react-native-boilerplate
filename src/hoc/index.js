export { default as withApp } from './withApp/withApp.hoc'
export { default as renderWhen } from './renderWhen/renderWhen.hoc'
export { default as withLoading } from './withLoading/withLoading.hoc'
