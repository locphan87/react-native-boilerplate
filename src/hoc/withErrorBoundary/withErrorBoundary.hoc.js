// @flow
import React from 'react'
import { compose, prop } from 'ramda'
import { isNonEmptyArray } from 'ramda-adjunct'

import { Wrapper, Text } from '../../components/Styled'
import renderWhen from '../renderWhen/renderWhen.hoc'

const ErrorComponent = () => (
  <Wrapper>
    <Text>Something went wrong, you can try to refetch again</Text>
  </Wrapper>
)
const hasError = compose(
  isNonEmptyArray,
  prop('errors')
)
const withErrorBoundary = renderWhen([
  {
    when: hasError,
    render: ErrorComponent
  }
])

export { ErrorComponent, hasError }
export default withErrorBoundary
