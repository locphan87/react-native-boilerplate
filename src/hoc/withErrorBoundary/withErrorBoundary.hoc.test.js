import { testSnapshots } from '../../utils/test.util'

import { ErrorComponent, hasError } from './withErrorBoundary.hoc'

const props = {
  errors: [
    {
      code: 'INVALID_INPUTS',
      message: 'Invalid field inputs'
    }
  ]
}

describe('HOC withErrorBoundary', () => {
  testSnapshots(ErrorComponent, [
    {
      props,
      description: 'basic render'
    }
  ])

  describe('hasError', () => {
    test('should return true when having errors', () => {
      expect(hasError({ errors: [{ code: 'NOT_FOUND' }] })).toEqual(true)
    })
    test('should return false when no errors', () => {
      expect(hasError({ errors: [] })).toEqual(false)
    })
  })
})
