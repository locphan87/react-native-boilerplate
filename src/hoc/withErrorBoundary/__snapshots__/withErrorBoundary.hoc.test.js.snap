// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOC withErrorBoundary snapshots basic render 1`] = `
<StyledWrapper>
  <StyledText>
    Something went wrong, you can try to refetch again
  </StyledText>
</StyledWrapper>
`;
